File: 2023/day_02/2.py
#!/usr/bin/env python3

import re


# games structure :
# {
#   1: [ { red: x, green: y,blue: z }, ...],
#   ...
# }
games = { }

with open('./input.txt') as file:
    lines = file.readlines()
    parsegame = "^Game ([0-9]+):(.*)$"
    parser = re.compile(parsegame)
    for line in lines:
        game, results = parser.match(line).groups()
        sequences = []
        for seq in results.split(";"): # sequence
            sequence = { "red": 0, "green": 0, "blue": 0}
            for color in seq.split(","): # ball ammount
                scorecolor = re.match(" ([0-9]+) (red|green|blue)", color).groups()
                sequence[scorecolor[1]] = int(scorecolor[0])
            sequences += [sequence]
        games[int(game)] = sequences

sum = 0
for game, plays in games.items():
    minrequired = { 'red': 0, 'green': 0, 'blue': 0}
    for play in plays:
        for color in minrequired:
            if play[color] > minrequired[color] : 
                minrequired[color] = play[color]
    power = 1
    for x in minrequired.values():
        power = power * x
    sum+=power
        
print(sum)

File: 2024/day_01/2.jl
#!/usr/bin/env julia 

lines = []

open("./input.txt") do file
  for line in eachline(file)
    push!(lines,line)
  end
end

cols = eachrow(
    reshape(
      vcat(map(
        line -> map(n -> parse(Int, n), split(line)), 
        lines)...),
      (2,:)))

ammount = Dict{Int,Int}()

map(x -> haskey(ammount,x) ? ammount[x] += 1 : ammount[x] = 1, cols[2])

total = 0
map(x -> if (haskey(ammount,x)) global total += ammount[x]*x end, cols[1])
println(total)
